### DISCUSSION TITLE
How to create dynamic fields in a JavaScript index?

### DISCUSSION THREAD

**Discussion by nickchampion** [ravendb/ravendb#14526] [General]
Hi,

Using the .NET client you used to be able to dynamically create index fields using `createField` I can't see a way of doing this with JavaScript indexes, is this still possible?

I'm using RavenDB Cloud on 5.3

Thanks

**Comment by ppekrol** [comment]
hi @nickchampion 

You can use `function createField(name, value, options)`. For example:

```
map('CreateFieldItems', function (i) {
                                    return {
                                        _: createField(i.FieldName, i.FieldValue, { indexing: 'Exact', storage: true, termVector: null })
                                    };
                                })
```

Hope that this helps.

**Reply by nickchampion** [reply]
Thanks @ppekrol but I don't seem to have that option, am I inheriting from the correct type?

<img width="878" alt="image" src="https://user-images.githubusercontent.com/3949005/178488946-fa321084-c59a-40ec-98f6-8cbd42643e81.png">

**Reply by ppekrol** [reply]
Looks like an issue with typings. @ml054 Can you take a look?

**Reply by ml054** [reply]
Looks like createField is missing in this.mapUtils()

Issue: https://issues.hibernatingrhinos.com/issue/RDBC-595/Expose-createField-function-in-maputils

**Reply by ml054** [reply]
@nickchampion as temporary workaround you can have createField function in context (even w/o implementation) and use that with parameters descrived above. Index definition calls toString() to on function so having in string createField should do the job.

**Reply by ml054** [reply]
Example: 
![image](https://user-images.githubusercontent.com/1250431/178562077-0eaf2fb4-0262-4608-ad12-838c0bd88eea.png)


Then in Studio it generates following defintion:
```
map('Items', doc => {
            return {
                name: doc.name,
                coordinates: createSpatialField(doc.lat, doc.lng),
                _: createField('test', 'test', { indexing: 'Exact', storage: true, termVector: null })
            };
        })
```

And nicely show terms:

![image](https://user-images.githubusercontent.com/1250431/178562214-0918802c-354e-417b-9c26-264c784c194f.png)

**Reply by nickchampion** [reply]
Thats great, thanks very much for the swift response @ml054 and @ppekrol

**Reply by ml054** [reply]
You are welcome - of course we fix that in node.js client and provide with strongly typed API for creating custom fields. :)

**Comment by nickchampion** [comment]
Not sure if I am missing something @ml054 but I am unable to create multiple custom fields using the approach above, this is my index

<img width="979" alt="image" src="https://user-images.githubusercontent.com/3949005/178578389-dea9cbc4-a99a-48d7-9ffc-74f578526dce.png">

and this is what I see in the terms tab of the index

<img width="1951" alt="image" src="https://user-images.githubusercontent.com/3949005/178578479-f36dfb99-c50e-4d34-aa17-39eb868cd7a6.png">

Should I be able to do what I am doing here to create multiple dynamic fields?

Thanks again

**Reply by ml054** [reply]
@nickchampion You can track progress here: https://issues.hibernatingrhinos.com/issue/RavenDB-18936/Creating-multiple-dynamic-fields-in-JS-index-doesnt-seem-to-work

**Reply by nickchampion** [reply]
Thanks @ml054

**Reply by nickchampion** [reply]
@ml054 would a potential workaround be to use an `AbstractIndexCreationTask` and create the map function in C#? we really need this feature and cant wait for the bug fix, any other suggestions for a workaround?

Thanks

**Reply by ayende** [reply]
As a workaround, you can emit the fields directly to the object.

```
this.map(storage =>
{
  var result = {};

  result["Dynamic"] = "foo";

 return result;
}
```

**Reply by nickchampion** [reply]
Thanks @ayende I'll give that a go

**Reply by ml054** [reply]
@nickchampion https://issues.hibernatingrhinos.com/issue/RavenDB-18936 was resolved and will be available in next stable release

**Reply by ml054** [reply]
@nickchampion This is fixed in 5.2.6
